File: Code/CSharp/Decorator/decorator.cs
public interface Train {
    string describe();
}

public class CargoTrain : Train {
    private int id;
    public CargoTrain(int id) {
        this.id = id;
    }

    public string describe() {
        return id + " is a cargo train. ";
    }
}

public class Modification : Train {
    private Train train;

    public Modification(Train train) {
        this.train = train;
    }

    public virtual string describe() {
        return train.describe();
    }
}

public class ColorModification : Modification {
    private string color;
    public ColorModification(Train train, string color) : base(train) {
        this.color = color;
    }

    public override string describe() {
        return base.describe() + "It is painted " + color + ". ";
    }
}

public class SilencerModification : Modification {
    public SilencerModification(Train train) : base(train) {
    }

    public override string describe() {
        return base.describe() + " When it drives it makes no sound. ";
    }
}

public class Tester {
    public static void Main(string[] args) {
        ColorModification classicTrain = new ColorModification(new CargoTrain(6), "red");
        SilencerModification stealthTrain = new SilencerModification(new ColorModification(new CargoTrain(7), "black"));
        System.Console.WriteLine(classicTrain.describe());
        System.Console.WriteLine(stealthTrain.describe());
    }
}

File: Code/Python/Decorator/decorator.py
class CargoTrain:
    def __init__(self, id):
        self.id = id

    def describe(self):
        return f"{self.id} is a cargo train. "

class Modification:
    def __init__(self, train):
        self.train = train

    def describe(self):
        return self.train.describe()

class ColorModification(Modification):
    def __init__(self, train, color):
        super().__init__(train)
        self.color = color
    def describe(self):
        return super().describe() + f"It is painted {self.color}. "

class SilencerModification(Modification):
    def describe(self):
        return super().describe() + f"When it drives it makes no sound. "

classicTrain = ColorModification(CargoTrain(6), "red")
stealthTrain = SilencerModification(ColorModification(CargoTrain(7), "black"))
print(classicTrain.describe())
print(stealthTrain.describe())